agent/engine/docker_image_manager.go

```diff
@@ -303,21 +309,145 @@ func (imageManager *dockerImageManager) removeUnusedImages(ctx context.Context)
 	imageManager.updateLock.Lock()
 	defer imageManager.updateLock.Unlock()
 
-	imageManager.imageStatesConsideredForDeletion = imageManager.imagesConsiderForDeletion(imageManager.getAllImageStates())
+	var ECSImagesDeletedTimes int
+	imageManager.imageStatesConsideredForDeletion = imageManager.imagesConsiderForDeletion(imageManager.getAllImageStates(), ctx)
 	for i := 0; i < imageManager.numImagesToDelete; i++ {
 		err := imageManager.removeLeastRecentlyUsedImage(ctx)
+		ECSImagesDeletedTimes = i + 1
 		if err != nil {
 			seelog.Infof("End of eligible images for deletion: %v; Still have %d image states being managed", err, len(imageManager.getAllImageStates()))
 			break
 		}
 	}
+	if imageManager.deleteNonECSImagesEnable {
+		// remove nonecs containers
+		imageManager.removeNonECSContainers(ctx)
+		// remove nonecs images
+		var NonECSImagesNumToDelete = imageManager.numImagesToDelete - ECSImagesDeletedTimes
+		imageManager.removeNonECSImages(NonECSImagesNumToDelete, ctx)
+	}
+}
+
+func (imageManager *dockerImageManager) removeNonECSContainers(ctx context.Context) {
+	nonECSContainerIds := imageManager.nonECSContainersIds(ctx)
+	var nonECSContainerRemoveAvailableIds []string
+	for _, id := range nonECSContainerIds {
+		response, _ := imageManager.client.InspectContainer(ctx, id, dockerclient.InspectContainerTimeout)
+		if response.State.Status == "exited" && time.Now().Sub(response.State.FinishedAt) > imageManager.taskCleanupWaitDuration {
+			nonECSContainerRemoveAvailableIds = append(nonECSContainerRemoveAvailableIds, id)
+		}
+	}
+	var numContainerAlreadyDeletePeriod = 0
+	for _, id := range nonECSContainerRemoveAvailableIds {
+		if numContainerAlreadyDeletePeriod == imageManager.numNonECSContainersToDelete {
+			break
+		}
+		err := imageManager.client.RemoveContainer(ctx, id, dockerclient.RemoveContainerTimeout)
+		if err == nil {
+			numContainerAlreadyDeletePeriod++
+		} else {
+			continue
+		}
+	}
+}
+
+func (imageManager *dockerImageManager) nonECSContainersIds(ctx context.Context) []string {
+	ecsContainerIds := imageManager.state.GetAllContainerIDs()
+	var allContainerIds []string
+	listContainersResponse := imageManager.client.ListContainers(ctx, true, dockerclient.ListContainersTimeout)
+	if listContainersResponse.Error != nil {
+		return []string{"listContainersResponse Error"}
+	}
+	for _, dockerID := range listContainersResponse.DockerIDs {
+		allContainerIds = append(allContainerIds, dockerID)
+	}
+	nonECSContainerIds := exclude(allContainerIds, ecsContainerIds)
+	return nonECSContainerIds
+}
+
+type ImageWithSize struct {
+	ImageName string
+	Size      int64
+}
+
+func (imageManager *dockerImageManager) removeNonECSImages(NonECSImagesNumToDelete int, ctx context.Context) {
+	var nonECSImageNames, nonECSImageExclusion = imageManager.nonECSImagesNameWithExclusionList(ctx)
+	nonECSImageNamesRemoveEligible := exclude(nonECSImageNames, nonECSImageExclusion)
```

Comment:
nit: the combination of **exclude()** and **nonECSImagesNameWithExclusionList()** actually introduces redundant iteration over imageList. That will not be a big issue for performance, however, maybe we can combine them together. Just have a helper function to check whether current image is in the exclusion list?